adapters/coap-vertx/src/main/java/org/eclipse/hono/adapter/coap/vertx/VertxBasedCoapAdapter.java

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.adapter.coap.vertx;
+
+import java.util.List;
+
+import org.eclipse.californium.core.CoapServer;
+import org.eclipse.californium.core.coap.CoAP.ResponseCode;
+import org.eclipse.californium.core.server.resources.CoapExchange;
+import org.eclipse.hono.adapter.coap.AbstractVertxBasedCoapAdapter;
+import org.eclipse.hono.adapter.coap.CoapAdapterProperties;
+import org.eclipse.hono.service.auth.device.Device;
+import org.eclipse.hono.util.Constants;
+import org.eclipse.hono.util.EventConstants;
+import org.eclipse.hono.util.TelemetryConstants;
+
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+
+/**
+ * A Vert.x based Hono protocol adapter for accessing Hono's Telemetry &amp; Event API using COAP.
+ */
+public final class VertxBasedCoapAdapter extends AbstractVertxBasedCoapAdapter<CoapAdapterProperties> {
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @return {@link Constants#PROTOCOL_ADAPTER_TYPE_COAP}
+     */
+    @Override
+    protected String getTypeName() {
+        return Constants.PROTOCOL_ADAPTER_TYPE_COAP;
+    }
+
+    /**
+     * Get extended device.
+     * 
+     * @param exchange coap exchange with URI.
+     * @param identityLevel path level to extract URI identity. e.g. for "/telemetry/tenant/authid" the level is 1.
+     * @param handler handler for determined extended device
+     */
+    public void getExtendedDevice(final CoapExchange exchange, final int identityLevel,
+            final Handler<ExtendedDevice> handler) {
+        if (identityLevel > 0) {
+            final List<String> path = exchange.getRequestOptions().getUriPath();
+            if (path.size() == identityLevel + 2) {
+                final Device device = new Device(path.get(identityLevel), path.get(identityLevel + 1));
```

Comment:
you may want to consider using the `org.eclipse.hono.util.ResourceIdentifier` class which provides semantic access to the URI's segments.

Reply:
I'm not sure, if you just want to exchange the extraction of the tenant and device id, or to replace the usage of `Device` or the usage of the `ExtendedDevice`. 

I introduced the usage of the `ExtendedDevice`, because it seems for me to be the "most straight forward" approach to pass the information to the abstract adapter, especially, if prinicpals are used to identify the "authorized device".

Or should it only be used to replace the usage of `Device` for the origin?
Or the extracting the origin device from the URL?

Reply:
No, I just wanted to suggest using `ResourceIdentifier.from(String)` in order to parse the URI and access the endpoint name, tenant and device IDs using `ResourceIdentifier`'s getters, instead of working on a bare String array ...

Reply:
Coap split already the uri path into several string parts. If you want to reuse the indexing, I can use the `ResourceIdentifier.fromPath(final String[] path)`. But this implementation seems not validating the path too much :-), so I'm not sure, if the validation should not implemented or external implemented, which would make it more fragile for changes.
My feeling, this  `ResourceIdentifier` is too generic to work her well.